### ISSUE TITLE
rank_test_score in BayesSearchCV is not working properly

### ISSUE THREAD

**Issue by xtosis** [scikit-optimize/scikit-optimize#604]
I'm using `BayesSearchCV` to do my hyper parameter tuning with `scoring` set to `'roc_auc'`. It correctly identifies the best `params` and `mean_test_score`  but the rankings...

> `mean_test_score` 0.7829 `rank_test_score` 1
> `mean_test_score` 0.7812 `rank_test_score` 1
> `mean_test_score` 0.7788 `rank_test_score` 2
> `mean_test_score` 0.7783 `rank_test_score` 3
> `mean_test_score` 0.7771 `rank_test_score` 4
> `mean_test_score` 0.7746 `rank_test_score` 2
> `mean_test_score` 0.7746 `rank_test_score` 2
> `mean_test_score` 0.7746 `rank_test_score` 2
> `mean_test_score` 0.7746 `rank_test_score` 2
> `mean_test_score` 0.7746 `rank_test_score` 2
> `mean_test_score` 0.7719 `rank_test_score` 5
> `mean_test_score` 0.7717 `rank_test_score` 6
> `mean_test_score` 0.7702 `rank_test_score` 7
> `mean_test_score` 0.7695 `rank_test_score` 7
> `mean_test_score` 0.7685 `rank_test_score` 8
> `mean_test_score` 0.7586 `rank_test_score` 1

Cloned your repo today and this is still happening

**Comment by betatim** [comment]
Do you have a code snippet that we can use to reproduce this?

**Comment by xtosis** [comment]
This is not the actual code that I am working on (that is a little bit long) but you can still see it happen with this as well for any model that you give it i.e. not just Random Forest Classifier:

```
from skopt import BayesSearchCV as BSCV
from skopt.space import Real, Categorical, Integer

RFC = RandomForestClassifier(max_features='sqrt', bootstrap=False)

par_rfc = {'n_estimators':Integer(14, 170), 
           'criterion':Categorical(['gini','entropy']), 
           'max_depth':Integer(1, 14), 
           'random_state':Categorical([5482, 46, 77, 2017, 2018, 25, 4])}

hps = BSCV(RFC, par_rfc, n_iter= 20, random_state=2017, n_jobs=8, cv=8, scoring='roc_auc', return_train_score=False, iid=False)

hps.fit(X_train, y_train)

res = pd.DataFrame(hps.cv_results_)
res = res.sort_values(by=['mean_test_score'], ascending=False, axis=0)
res.reset_index(inplace = True, drop=True)
a = res[['mean_test_score', 'std_test_score']]
res.drop(['mean_test_score', 'std_test_score', 'params', 'mean_fit_time', 'mean_score_time', 'std_fit_time', 'std_score_time'], axis = 1, inplace = True)

res = pd.concat([a, res], axis=1)
print("best_score_: {:.4f}\nbest_params_: {}".format(hps.best_score_, hps.best_params_))
display(res.head(20))
```

**Output**
```
best_score_: 0.7859
best_params_: {'n_estimators': 143, 'random_state': 77, 'criterion': 'entropy', 'max_depth': 7}
```

  | mean_test_score | std_test_score | param_criterion | param_max_depth | param_n_estimators | param_random_state | rank_test_score | split0_test_score | split1_test_score | split2_test_score | split3_test_score | split4_test_score | split5_test_score | split6_test_score | split7_test_score
| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 | 0.7859 | 0.0476 | entropy | 7 | 143 | 77 | 1 | 0.7994 | 0.8193 | 0.7971 | 0.8084 | 0.7684 | 0.7973 | 0.8291 | 0.6681
1 | 0.7829 | 0.0453 | entropy | 9 | 163 | 77 | 2 | 0.7947 | 0.8101 | 0.7695 | 0.8231 | 0.7700 | 0.8044 | 0.8176 | 0.6736
2 | 0.7775 | 0.0288 | entropy | 11 | 158 | 2017 | 1 | 0.7894 | 0.8009 | 0.7652 | 0.8014 | 0.7754 | 0.7846 | 0.7951 | 0.7077
3 | 0.7774 | 0.0449 | entropy | 3 | 162 | 4 | 3 | 0.8009 | 0.7992 | 0.7841 | 0.8041 | 0.7511 | 0.7841 | 0.8247 | 0.6709
4 | 0.7762 | 0.0449 | gini | 4 | 42 | 4 | 4 | 0.7821 | 0.8074 | 0.7863 | 0.8079 | 0.7435 | 0.7978 | 0.8137 | 0.6709
5 | 0.7759 | 0.0413 | entropy | 5 | 80 | 2018 | 5 | 0.8025 | 0.8111 | 0.7700 | 0.8014 | 0.7587 | 0.7890 | 0.7989 | 0.6758
6 | 0.7719 | 0.0392 | entropy | 14 | 170 | 5482 | 2 | 0.7884 | 0.8220 | 0.7348 | 0.7965 | 0.7792 | 0.7692 | 0.7959 | 0.6890
7 | 0.7719 | 0.0392 | entropy | 14 | 170 | 5482 | 2 | 0.7884 | 0.8220 | 0.7348 | 0.7965 | 0.7792 | 0.7692 | 0.7959 | 0.6890
8 | 0.7719 | 0.0392 | entropy | 14 | 170 | 5482 | 2 | 0.7884 | 0.8220 | 0.7348 | 0.7965 | 0.7792 | 0.7692 | 0.7959 | 0.6890
9 | 0.7719 | 0.0392 | entropy | 14 | 170 | 5482 | 2 | 0.7884 | 0.8220 | 0.7348 | 0.7965 | 0.7792 | 0.7692 | 0.7959 | 0.6890
10 | 0.7719 | 0.0392 | entropy | 14 | 170 | 5482 | 2 | 0.7884 | 0.8220 | 0.7348 | 0.7965 | 0.7792 | 0.7692 | 0.7959 | 0.6890
11 | 0.7707 | 0.0441 | entropy | 2 | 108 | 4 | 7 | 0.8017 | 0.7892 | 0.7716 | 0.7976 | 0.7427 | 0.7747 | 0.8192 | 0.6690
12 | 0.7688 | 0.0389 | gini | 12 | 50 | 25 | 6 | 0.7727 | 0.8122 | 0.7665 | 0.7933 | 0.7722 | 0.8000 | 0.7566 | 0.6769
13 | 0.7679 | 0.0390 | gini | 12 | 34 | 2018 | 7 | 0.7717 | 0.8166 | 0.7814 | 0.8084 | 0.7519 | 0.7643 | 0.7690 | 0.6802
14 | 0.7633 | 0.0338 | gini | 14 | 14 | 46 | 1 | 0.7604 | 0.8141 | 0.7189 | 0.8049 | 0.7505 | 0.7404 | 0.7912 | 0.7261
15 | 0.7603 | 0.0534 | entropy | 14 | 14 | 5482 | 8 | 0.7414 | 0.8293 | 0.7476 | 0.8139 | 0.7568 | 0.7508 | 0.7975 | 0.6456
16 | 0.7597 | 0.0386 | gini | 1 | 170 | 46 | 2 | 0.7905 | 0.7687 | 0.7738 | 0.7827 | 0.7394 | 0.7577 | 0.7967 | 0.6681
17 | 0.7569 | 0.0456 | entropy | 1 | 156 | 77 | 8 | 0.7874 | 0.7784 | 0.7825 | 0.7762 | 0.7357 | 0.7426 | 0.8025 | 0.6497
18 | 0.7560 | 0.0400 | entropy | 1 | 170 | 46 | 3 | 0.7751 | 0.7784 | 0.7735 | 0.7863 | 0.7367 | 0.7319 | 0.7992 | 0.6670
19 | 0.7309 | 0.0376 | entropy | 1 | 14 | 46 | 4 | 0.6852 | 0.7627 | 0.7462 | 0.7722 | 0.7384 | 0.6909 | 0.7736 | 0.6783

**Comment by betatim** [comment]
Thank you for the snippet! Will have a look to see where thing are going wrong. If you get there first let us know.

**Comment by chris-code** [comment]
This bug seems to stem from what is mentioned here:

https://github.com/scikit-optimize/scikit-optimize/blob/1c4c0f12ad8c1fe0a33542108fc0e55164138f9d/skopt/searchcv.py#L544

so when the rank is computed here:
https://github.com/scikit-optimize/scikit-optimize/blob/1c4c0f12ad8c1fe0a33542108fc0e55164138f9d/skopt/searchcv.py#L437-L438

array_means contains the scores for the currently selected set of n_points, rather than all points that have been evaluated so far. That also means that all ranks are 1 if one sets n_points=1.

Since this has not moved since January, would you be interested in a PR?

**Comment by holgern** [comment]
Fix be PR #832
